src/document-symbols.jl

```diff
@@ -0,0 +1,313 @@
+            end
+        end
+    elseif k === K"struct"
+        binfos, ctx = lower_and_get_bindings(ex)
+        struct_children = DocumentSymbol[]
+
+        type_param_bindings = filter(binfo -> binfo.kind === :static_parameter, binfos)
+        for tpb in type_param_bindings
+            tpb_symbol = _DocumentSymbol(tpb.name,
+                                         JL.binding_ex(ctx, tpb.id),
+                                         SymbolKind.TypeParameter)
+            push!(struct_children, tpb_symbol)
+        end
+        field_bindings = filter(binfo -> binfo.kind === :argument, binfos)
+        for fb in field_bindings
+            fb_symbol = _DocumentSymbol(fb.name,
+                                        JL.binding_ex(ctx, fb.id),
+                                        SymbolKind.Field)
+            push!(struct_children, fb_symbol)
+        end
+
+        struct_name_ex = children(ex)[1]
+        if kind(struct_name_ex) === K"curly"
+            struct_name_ex = children(struct_name_ex)[1]
+        end
+        struct_name = struct_name_ex.name_val
+        struct_symbol = _DocumentSymbol(struct_name,
+                                        ex,
+                                        SymbolKind.Struct,
+                                        struct_children)
+        push!(symbols, struct_symbol)
+    elseif k === K"::"
+        if JS.is_prefix_op_call(ex)
+            type_symbol = _DocumentSymbol(get_type_name(ex[1]),
+                                          ex[1],
+                                          SymbolKind.Struct)
+            push!(symbols, type_symbol)
+        else
+            get_symbols!(ex[1], symbols)
+            type_symbol = _DocumentSymbol(get_type_name(ex[2]),
+                                          ex[2],
+                                          SymbolKind.Struct)
+            push!(symbols, type_symbol)
+        end
+    elseif k === K"if" || k === K"block"
+        for c in children(ex)
+            get_symbols!(c, symbols)
+        end
+    elseif k === K"for" || k === K"while" || k === K"="
+        get_symbols!(ex[1], symbols)
+        get_symbols!(ex[2], symbols)
+    elseif k === K"const" || k === K"return"
+        get_symbols!(ex[1], symbols)
+    else
+        binfos, ctx = lower_and_get_bindings(ex)
+        for b in binfos
+            symbol = _DocumentSymbol(b.name,
+                                     JL.binding_ex(ctx, b.id),
+                                     SymbolKind.Variable)
+            push!(symbols, symbol)
+        end
+    end
+
+    return symbols
+end
+
+# Symbols search utils
+# --------------------
+
+"""
+Get the bindings from a `SyntaxTree` lowered up to scope resolution. Filter out internal
+bindings.
+"""
+function lower_and_get_bindings(st0::JL.SyntaxTree)
+    # Lower the syntax tree.
+    ctx1, st1 = JL.expand_forms_1(Module(), st0)
+    ctx2, st2 = JL.expand_forms_2(ctx1, st1)
+    ctx3, _ = JL.resolve_scopes(ctx2, st2)
```

Comment:
Lowering in general can produce arbitrary errors since it may include execution of user code, so we really need to wrap it into `try`/`catch` and handle the lowering failure somehow. See completions.jl for how it should be done. Reuse the lowering routine if possible.